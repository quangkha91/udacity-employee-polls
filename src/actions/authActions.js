export const LOGIN_SUCCESS = "LOGIN_SUCCESS";
export const LOGOUT = "LOGOUT";
export const LOGIN_ERROR = "LOGIN_ERROR";

export const loginSuccess = (user) => ({
  type: LOGIN_SUCCESS,
  payload: user,
});

export const loginError = (state) => ({
  type: LOGIN_ERROR,
  payload: state
});

export const logout = () => ({
  type: LOGOUT,
});

export const login = (username, password) => (dispatch, getState) => {
  const { users } = getState();
  const user = Object.values(users.data).find(
    (user) => user.id === username && user.password === password
  );
  if (!!user) {
    loginError(undefined)
    return dispatch(loginSuccess(user));
  }
  // Handle login failure
  return dispatch(loginError("Login failed. Please check your credentials."));
};
