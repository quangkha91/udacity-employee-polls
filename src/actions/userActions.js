export const INITIAL_USERS = "INITIAL_USERS";
export const USER_ANSWER_VOTE = "USER_ANSWER_VOTE";
export const USER_CREATE_QUESTION = "USER_CREATE_QUESTION";

export function updateAnswerUser(loginUser, questionId, answer) {
  return {
    type: USER_ANSWER_VOTE,
    payload: {
      userId: loginUser,
      qid: questionId,
      answer,
    },
  };
}

export function createQuestionUser( userId, questionId ) {
  return {
    type: USER_CREATE_QUESTION,
    payload: {
      userId,
      qid: questionId,
    },
  };
}
