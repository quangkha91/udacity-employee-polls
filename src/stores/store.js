import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authReducer';
import userReducer from '../reducers/userReducer';
import questionReducer from '../reducers/questionReducer';

const store = configureStore({
  reducer: {
    auth: authReducer,
    users: userReducer,
    questions: questionReducer
  },
});

export default store;