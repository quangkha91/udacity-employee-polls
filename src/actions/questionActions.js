import { createQuestion, updateQuestionAnswer } from "../util/data";
import { createQuestionUser, updateAnswerUser } from "./userActions";

export const INITIAL_QUESTIONS = "INITIAL_QUESTIONS";
export const CREATE_QUESTION = "CREATE_QUESTION";
export const ANSWER_QUESTION = "ANSWER_QUESTION";

function createNewQuestion(question) {
  return {
    type: CREATE_QUESTION,
    payload: {
      question,
    },
  };
}

function answerQuestion(userId, questionId, answer) {
  return {
    type: ANSWER_QUESTION,
    payload: {
      userId,
      qid: questionId,
      answer,
    },
  };
}

export function handleCreateQuestion(firstOption, secondOption) {
  return (dispatch, getState) => {
    const { auth } = getState();

    return createQuestion(firstOption, secondOption, auth.user).then(
      (question) => {
        dispatch(createNewQuestion(question));
        dispatch(createQuestionUser(auth.user.id, question.id));
      }
    );
  };
}

export function handleUpdateAnswer(questionId, answer) {
  return async (dispatch, getState) => {
    const { auth } = getState();
    await updateQuestionAnswer(auth.user.id, questionId, answer).then(() => {
      dispatch(answerQuestion(auth.user.id, questionId, answer));
      dispatch(updateAnswerUser(auth.user.id, questionId, answer));
    });
  };
}
