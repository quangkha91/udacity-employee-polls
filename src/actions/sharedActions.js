import { INITIAL_USERS } from "./userActions";
import { INITIAL_QUESTIONS } from "./questionActions";
import { getInitialData } from "../util/data";

const initialUsers = (users) => {
  return {
    type: INITIAL_USERS,
    payload: users,
  };
};

const initialQuestions = (questions) => {
  return {
    type: INITIAL_QUESTIONS,
    payload: questions,
  };
};

export function handleInitialData() {
  return async (dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(initialUsers(users));
      dispatch(initialQuestions(questions));
    });
  };
}
