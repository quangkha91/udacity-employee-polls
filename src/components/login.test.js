import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import store from "../stores/store";
import Login from "./login";
import { loginError } from "../actions/authActions";

describe("Login screen", () => {
  it("Render a snapshot for Login renderer", async () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("should show error when user login false", async () => {
    await store.dispatch(
      loginError("Login failed. Please check your credentials.")
    );

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    
    const errorText = component.getByTestId("login-error");
    console.log(errorText);
    expect(errorText).toBeInTheDocument();
    expect(errorText.textContent).toBe('Login failed. Please check your credentials.');
  });
});
