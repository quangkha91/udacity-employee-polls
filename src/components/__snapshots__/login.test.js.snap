// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen Render a snapshot for Login renderer 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
        data-testid="login-page"
      >
        <form>
          <h2
            class="title"
          >
            Login
          </h2>
          <hr
            class="title-line"
          />
          <div
            class="login-form-group"
          >
            <label
              for="username"
            >
              Username:
            </label>
            <input
              data-testid="username"
              id="username"
              name="username"
              required=""
              type="text"
              value="user"
            />
          </div>
          <div
            class="login-form-group"
          >
            <label
              for="password"
            >
              Password:
            </label>
            <input
              data-testid="password"
              id="password"
              name="password"
              required=""
              type="password"
              value="password"
            />
          </div>
          <div
            class="login-form-group"
          >
            <button
              data-testid="submit"
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
      data-testid="login-page"
    >
      <form>
        <h2
          class="title"
        >
          Login
        </h2>
        <hr
          class="title-line"
        />
        <div
          class="login-form-group"
        >
          <label
            for="username"
          >
            Username:
          </label>
          <input
            data-testid="username"
            id="username"
            name="username"
            required=""
            type="text"
            value="user"
          />
        </div>
        <div
          class="login-form-group"
        >
          <label
            for="password"
          >
            Password:
          </label>
          <input
            data-testid="password"
            id="password"
            name="password"
            required=""
            type="password"
            value="password"
          />
        </div>
        <div
          class="login-form-group"
        >
          <button
            data-testid="submit"
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
